File: lcode100-199/ex121/deleteDuplicates.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        deleted = []
        added = []
        cur = head
        while cur != None:
            value = cur.val
            if value not in deleted:
                if value not in added:
                    added.append(value)
                else:
                    added.remove(value)
                    deleted.append(value)
                cur = cur.next
            else:
                cur = cur.next
                continue
            
        
        cur = ListNode(-1)
        newHead = cur
        for i in added:
            newHead.next = ListNode(i)
            newHead = newHead.next
        return cur.next
        
        

"""
 1->2->3->3->4->4->5
 给定一个排序链表，删除所有含有重复数字的节点，只保留原始链表中 没有重复出现 的数字。
"""

File: lcode100-199/ex166/deleteDuplicates.go
// Copyright 2019 The darwin Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package main

import "fmt"

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteDuplicates(head *ListNode) *ListNode {
	// special case
	if head == nil || head.Next == nil {
		return head
	}
	before, after := head, head.Next

	for after != nil {
		for after != nil && before.Val == after.Val {
			after = after.Next
		}
		before.Next = after
		before = before.Next
		if after == nil {
			break
		}
		after = after.Next
	}
	return head
}

func main() {
	head := &ListNode{
		2,
		&ListNode{
			2,
			&ListNode{
				2,
				nil,
			},
		},
	}
	head2 := deleteDuplicates(head)
	for head2 != nil {
		fmt.Println(head2.Val)
		head2 = head2.Next
	}
}
